istioctl/pkg/auth/upgrade.go

```diff
@@ -220,14 +223,27 @@ func (ug *Upgrader) getAndAddRoleNameToWorkloadLabelMapping(roleName, namespace
 			return nil, err
 		}
 	}
-	for _, serviceWorkloadLabels := range ug.RoleNameToWorkloadLabels[roleName] {
-		for serviceName, labels := range serviceWorkloadLabels {
-			workloadLabels[serviceName] = labels
+	// A ServiceRole can have a list of rules. We need to go through all services in all rules for a ServiceRole.
+	// A serviceWorkloadLabels is the mapping from pod selector (for example app: productpage).
+	for _, serviceWorkloadLabels := range ug.NamespaceToRoleNameToWorkloadLabels[namespace][roleName] {
+		// For the example above, labelKey is app, labelValue is productpage.
+		for labelKey, labelValue := range serviceWorkloadLabels {
```

Comment:
I don't think so. The reason being, if you have a ServiceRole with 2 rules, one with service `foo` and one with service `bar`. If `foo` uses `app1: foo` label, and `bar` uses `app2: bar` label, `workloadLabels` will be just one of `app1: foo` or `app2: bar`. What we want is 
```
workloadLabels = {
  app1: foo
  app2: bar
}
```
If both `app1` and `app2` (the keys) are the same, e.g. `app`, then we just use the second one. I'm not sure if this is a safe approach, but since the labels from the API is `map[string]string`. I think we can just change label API to a list of strings?

Reply:
Discussed offline. We will break n rules to n AuthorizationPolicies :)